### PR TITLE
feat(image_optimizer_default_settings): add Image Optimizer default settings API

### PR DESCRIPTION
**Pull request by daboross**
Includes an update to go-fastly 9.4.0.

### CODE REVIEW

File: fastly/block_fastly_service_image_optimizer_default_settings.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+			apiInput.JpegType = &jpegType
+		case "jpeg_quality":
+			jpegQuality := value.(int)
+			apiInput.JpegQuality = &jpegQuality
+		case "upscale":
+			upscale := value.(bool)
+			apiInput.Upscale = &upscale
+		case "allow_video":
+			allowVideo := value.(bool)
+			apiInput.AllowVideo = &allowVideo
```
Comment (Integralist) [inline fastly/block_fastly_service_image_optimizer_default_settings.go]: Everywhere you've created a variable for the sake of taking its address you can instead use `gofastly.ToPointer(value.(whatever))`.

**Review by Integralist — COMMENTED** [review]
Just one small comment for the moment, but it's looking good. Let me know when you're ready for a full review.

File: fastly/block_fastly_service_image_optimizer_default_settings.go (hunk 1 above)
Comment (daboross) [reply fastly/block_fastly_service_image_optimizer_default_settings.go]: Oh awesome! That's much better haha.

**Comment by daboross** [comment]
@Integralist I'd request a full review whenever you've got the time!

I'll leave this as a draft until the api-documentation & go-fastly PRs are merged, but I've finally gotten this working.

Also, my apologies, I made some modifications to the go-fastly PR to better test it & better support the terraform use case. Probably should have waited until I had everything working from one end to the other to submit PRs, but well you live & learn.

I'll clean up / squash the commit messages - they're a bit of a mess - but the code itself should be good.

**Comment by Integralist** [comment]
👋🏻 @daboross apologies I've been stacked up with other work and I didn't realise it's been 2 weeks since you last commented. So I'll try to take a look at this PR tomorrow morning for you.

**Comment by daboross** [comment]
:wave:

It's no problem! I also could have commented something: after finalizing the api-documentation PR, I went and cleaned up the go-fastly PR, and was starting to clean up this one up when I got sidetracked by a missing consideration in the original API.

I'll push those commits so tomorrow morning you can check it out, and leave a comment where I have one part untested because of a change needed to the API itself.

File: fastly/block_fastly_service_image_optimizer_default_settings.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	log.Printf("[DEBUG] Calling Image Optimizer default settings update API with parameters: %+v", apiInput)
+
+	if _, err := conn.UpdateImageOptimizerDefaultSettings(&apiInput); err != nil {
+		// inspect the error type for a title that has a message indicating the user cannot call the API because they do not have Image Optimizer
+		// enabled. For these users we want to skip the error so that we can allow them to clean up their Terraform state. (also, because the Image Optimizer
+		// default settings for services with Image Optimizer are effectively cleared by disabling Image Optimizer.)
+		if he, ok := err.(*gofastly.HTTPError); ok {
+			if he.StatusCode == http.StatusBadRequest {
+				for _, e := range he.Errors {
+					if strings.Contains(e.Detail, "Image Optimizer is not enabled on this service") {
```
Comment (daboross) [inline fastly/block_fastly_service_image_optimizer_default_settings.go:324]: This is the untested bit! When editing the docs for this PR, I was thinking about whether we needed a warning for this API requiring customers to be entitled, and realized we never thought about nor implemented any checks for that.

So I'm working on a PR to add a requirement that Image Optimizer is enabled on the service, and I'm waiting to merge the api-documentation PR (and to un-draft these) until that's ready.

File: fastly/block_fastly_service_image_optimizer_default_settings_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		ProviderFactories: testAccProviders,
+		CheckDestroy:      testAccCheckServiceVCLDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckServiceExists("fastly_service_vcl.foo", &service),
+					resource.TestCheckResourceAttr("fastly_service_vcl.foo", "name", serviceName),
+					resource.TestCheckResourceAttr("fastly_service_vcl.foo", "backend.#", "1"),
+					testAccCheckFastlyServiceVCLBackendAttributes(&service, []*gofastly.Backend{&b1}),
```
Comment (Integralist) [inline fastly/block_fastly_service_image_optimizer_default_settings_test.go]: We should have some test assertions for validating the `image_optimizer_default_settings` block was created correctly in the TF state, and also add an 'update' flow so we can be sure it updates correctly too.

**Review by Integralist — COMMENTED** [review]
LGTM, just one comment re: tests...

File: fastly/block_fastly_service_image_optimizer_default_settings_test.go (hunk 3 above)
Comment (daboross) [reply fastly/block_fastly_service_image_optimizer_default_settings_test.go]: Thanks for commenting on this! I've added a validation check, and a second step for checking updating.

**Comment by daboross** [comment]
Going to rebase to fix merge conflicts, keeping all commits as is otherwise.

**Review by Integralist — APPROVED** [review]
LGTM. Mark the PR 'ready for review' and I'll give it a final going over.

File: fastly/block_fastly_service_image_optimizer_default_settings.go (hunk 2 above)
Comment (daboross) [reply fastly/block_fastly_service_image_optimizer_default_settings.go:324]: I've now tested that this works.

**Comment by daboross** [comment]
This should be ready for a full review! I've included an update to the released go-fastly 9.4.0.

**Review by Integralist — APPROVED** [review]

**Pull request merged**